File: python/vcp_gateway/command.py
"""Command execution enforcement — sandbox paths + the no-shell executor (§28).

The Gateway is the enforcing boundary for ``command`` capabilities (§28.2,
§28.5). This module:

* :func:`check_command_paths` — refuses a path parameter that escapes the
  ``sandbox.filesystem`` allowlist, by absolute path or by ``..`` traversal,
  with ``SANDBOX_VIOLATION`` (§28.2). Paths are normalized with POSIX
  semantics so ``/work/../etc/passwd`` cannot masquerade as in-scope.
* :func:`run_command` — the real executor. It runs a resolved argv array via
  ``subprocess.run([...], shell=False, ...)``. ``shell`` is **always** false:
  there is no ``/bin/sh -c`` / ``cmd /c`` / PowerShell, so a parameter such as
  ``"; rm -rf / #"`` is delivered to the program as one literal argv element and
  never interpreted as a new command (§28.1). A non-zero exit is a *result*, not
  a silent failure (§28.6).

Command output (``stdout`` / ``stderr``) is labelled ``untrusted_tool_result``
(§12, §28.5); use :func:`vcp_gateway.taint.authority_decision` to enforce that
such output can never authorize the next command
(``AUTHORITY_FROM_TAINTED_DATA``). That rule is reused from the existing taint
engine; nothing command-specific is needed to honor it.
"""

from __future__ import annotations

import posixpath
import subprocess
import sys
from dataclasses import dataclass
from typing import Any, Iterable, List, Mapping, Optional, Sequence

from vcp_sdk import reason_codes as rc
from vcp_sdk.command import is_param_hole

__all__ = [
    "COMMAND_OUTPUT_LABEL",
    "check_command_paths",
    "run_command",
    "CommandResult",
]

# The taint label every command's stdout/stderr carries (§12, §28.5). Importing
# this here keeps the §28.5 rule discoverable from the command module; the
# decision itself is made by vcp_gateway.taint.authority_decision.
COMMAND_OUTPUT_LABEL = "untrusted_tool_result"


def _normalize(path: str) -> str:
    """Normalize a POSIX-style path: collapse ``.``/``..`` without touching disk.

    We use :mod:`posixpath` (not :mod:`os.path`) so the check is deterministic
    across host OSes and matches the ``/work`` style allowlists in the vectors —
    a Windows test must still resolve ``/work/../etc/passwd`` the POSIX way.
    Absolute paths are normalized as-is; relative paths are normalized but kept
    relative so traversal above the root is detectable.
    """
    return posixpath.normpath(path)


def _is_within(candidate: str, root: str) -> bool:
    """True if normalized ``candidate`` is inside (or equal to) normalized ``root``."""
    c = _normalize(candidate)
    r = _normalize(root)
    if c == r:
        return True
    # Compare on path-component boundaries so /work does not match /workspace.
    prefix = r if r.endswith("/") else r + "/"
    return c.startswith(prefix)


def _iter_path_values(
    params: Optional[Mapping[str, Any]],
    argv: Optional[Sequence[str]],
    argv_template: Optional[Sequence[Any]],
) -> Iterable[str]:
    """Yield the candidate path strings to check.

    Two modes:

    * Given ``params`` (+ optional ``argv_template`` to know which params are
      paths): yield path-typed param values. A param is path-typed when its
      template hole carries ``schema.vcp_kind == "path"``; absent a template we
      conservatively treat every string param as a candidate path.
    * Given a resolved ``argv`` array with no params: yield every element that
      looks like a filesystem path (starts with ``/`` or contains ``..``).
    """
    if params is not None:
        path_params: Optional[set[str]] = None
        if argv_template is not None:
            path_params = set()
            for tok in argv_template:
                if is_param_hole(tok):
                    schema = tok.get("schema", {})
                    if schema.get("vcp_kind") == "path":
                        path_params.add(tok["param"])
        for key, value in params.items():
            if not isinstance(value, str):
                continue
            if path_params is None or key in path_params:
                yield value
        return
    if argv is not None:
        for element in argv:
            if isinstance(element, str) and (
                element.startswith("/") or ".." in element.split("/")
            ):
                yield element


def check_command_paths(
    params: Optional[Mapping[str, Any]] = None,
    sandbox_filesystem: Any = "none",
    *,
    argv: Optional[Sequence[str]] = None,
    argv_template: Optional[Sequence[Any]] = None,
) -> dict:
    """Refuse path params that escape ``sandbox.filesystem`` (§28.2).

    Returns ``{"decision": "allow", "reason_code": "OK"}`` when every path
    parameter resolves inside the allowlist, or
    ``{"decision": "deny", "reason_code": "SANDBOX_VIOLATION"}`` for the first
    value that escapes — either an absolute path outside the allowlist
    (``/home/user/.ssh/id_rsa``) or a relative ``..`` traversal that climbs out
    (``/work/../etc/passwd``). Paths are normalized with POSIX semantics so the
    traversal cannot hide.

    Pass either ``params`` (with an optional ``argv_template`` to identify which
    params are ``vcp_kind: path``) or a fully-resolved ``argv`` array.

    ``sandbox_filesystem`` is the manifest's ``sandbox.filesystem``: ``"none"``
    (no filesystem access — any path is a violation) or an allowlist array.
    """
    if sandbox_filesystem == "none":
        allowlist: List[str] = []
    elif isinstance(sandbox_filesystem, str):
        allowlist = [sandbox_filesystem]
    else:
        allowlist = list(sandbox_filesystem)

    for value in _iter_path_values(params, argv, argv_template):
        if not any(_is_within(value, root) for root in allowlist):
            return {
                "decision": "deny",
                "reason_code": rc.SANDBOX_VIOLATION,
                "remediation": {
                    "message": (
                        "path parameter escapes the sandbox.filesystem allowlist"
                    ),
                    "path": _normalize(value),
                    "allowed": allowlist,
                },
            }
    return {"decision": "allow", "reason_code": rc.OK}


@dataclass(frozen=True)
class CommandResult:
    """The result of a real, no-shell command execution (§28.6).

    A non-zero ``exit_code`` is a result, not a silent failure; ``stdout`` and
    ``stderr`` are labelled ``untrusted_tool_result`` (§28.5).
    """

    argv: List[str]
    exit_code: int
    stdout: str
    stderr: str
    shell: bool  # ALWAYS False — proves no shell was used.
    output_label: str = COMMAND_OUTPUT_LABEL

    def as_dict(self) -> dict:
        return {
            "argv": list(self.argv),
            "exit_code": self.exit_code,
            "stdout": self.stdout,
            "stderr": self.stderr,
            "shell": self.shell,
            "output_label": self.output_label,
        }


def run_command(
    argv: Sequence[str],
    *,
    cwd: Optional[str] = None,
    env: Optional[Mapping[str, str]] = None,
    timeout: Optional[float] = None,
) -> CommandResult:
    """Execute a resolved argv array directly — ``shell=False`` ALWAYS (§28.1).

    The Gateway MUST exec ``argv[0]`` with the argv **array**; it MUST NOT pass
    the command to a shell and MUST NOT perform shell interpolation, globbing,
    quoting, or word-splitting. A parameter such as ``"; rm -rf / #"`` is
    therefore delivered to the program as a single literal argv element and is
    never interpreted as a new command. ``env`` defaults to *no inherited
    environment* (§28.2) — pass an explicit map to supply broker-provided values.
    """
    if not argv:
        raise ValueError("argv must contain at least the binary")
    completed = subprocess.run(  # noqa: S603 - shell=False by construction (§28.1)
        list(argv),
        shell=False,  # NEVER True. This is the §28.1 invariant.
        capture_output=True,
        text=True,
        cwd=cwd,
        env=dict(env) if env is not None else {},
        timeout=timeout,
    )
    return CommandResult(
        argv=list(argv),
        exit_code=completed.returncode,
        stdout=completed.stdout,
        stderr=completed.stderr,
        shell=False,
    )

File: rust/crates/vcp-gateway/src/command.rs
//! Command / CLI capabilities — Gateway-side enforcement (§28).
//!
//! The SDK ([`vcp_sdk::command`]) owns the argv model, identity, and the command
//! bridge. The Gateway owns the parts that require *authority*:
//!
//! - [`check_command_paths`]: a sandbox path check (§28.2, extends §14). A path
//!   parameter that resolves outside the `sandbox.filesystem` allowlist — whether
//!   by an absolute path elsewhere, or a relative `..` escape — is refused with
//!   `SANDBOX_VIOLATION` (security test 21).
//! - [`command_authority`]: the taint rule applied to command output (§28.5): a
//!   command's `stdout`/`stderr` is `untrusted_tool_result` and can never
//!   authorize a command (`AUTHORITY_FROM_TAINTED_DATA`, security test 20-adjacent
//!   / taint case). Delegates to the existing [`crate::taint`] engine.
//! - [`run_argv`]: the real executor. It builds a [`std::process::Command`] from
//!   `binary` + the resolved argv array and **never** wraps it in a shell. This is
//!   the structural CWE-78 defense (§28.1): a metacharacter value is one literal
//!   argv element handed to the program, not a new command.

use crate::grant::Decision;
use crate::reason::ReasonCode;
use crate::taint;

/// Normalize a path into its logical components, resolving `.` and `..` lexically
/// (no filesystem access — we judge the *declared* path, not what is on disk).
/// A `..` that would escape above the root is preserved as a leading `..` so the
/// caller can detect an escape. Works for POSIX-style `/`-separated paths, which
/// is what the §28 vectors use.
fn normalize(path: &str) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    for seg in path.split('/') {
        match seg {
            "" | "." => {}
            ".." => {
                // Pop a real segment if there is one to pop; otherwise keep the
                // `..` so an escape above the root remains visible.
                if matches!(out.last().map(String::as_str), Some(s) if s != "..") {
                    out.pop();
                } else {
                    out.push("..".to_string());
                }
            }
            other => out.push(other.to_string()),
        }
    }
    out
}

/// Is `path` absolute (POSIX leading `/`)? The vectors are POSIX paths.
fn is_absolute(path: &str) -> bool {
    path.starts_with('/')
}

/// True iff the normalized `candidate` lies within the normalized `root`
/// (root itself, or any descendant). Both are treated as absolute roots.
fn within(root: &[String], candidate: &[String]) -> bool {
    // An escape (leading `..`) can never be within an absolute root.
    if candidate.first().map(String::as_str) == Some("..") {
        return false;
    }
    if candidate.len() < root.len() {
        return false;
    }
    root.iter().zip(candidate.iter()).all(|(r, c)| r == c)
}

/// Sandbox path check for a command parameter (§28.2). Given the resolved value
/// of a path-typed parameter and the `sandbox.filesystem` allowlist, decide
/// whether the path stays inside the sandbox.
///
/// Refused with `SANDBOX_VIOLATION` when the path resolves outside every
/// allowlisted root — either an absolute path elsewhere
/// (`/home/user/.ssh/id_rsa`) or a relative `..` escape
/// (`/work/../etc/passwd`), both normalized before comparison. A relative path
/// is anchored to each root before checking, so it cannot reach outside it.
///
/// `allowlist` is the `sandbox.filesystem` value: `"none"` (deny all paths) or an
/// array of allowed roots. A `"none"` sandbox refuses any path parameter.
pub fn check_command_paths(path: &str, allowlist: &[String]) -> (Decision, ReasonCode) {
    // Empty allowlist (or `none`) ⇒ no filesystem access permitted.
    if allowlist.is_empty() {
        return (Decision::Deny, ReasonCode::SandboxViolation);
    }

    let candidate_abs = if is_absolute(path) {
        normalize(path)
    } else {
        // A relative path is only meaningful when anchored. We check it against
        // each allowed root by anchoring it there; if it escapes the root via
        // `..` the normalized form retains a leading `..` and `within` fails.
        // Try every root.
        for root in allowlist {
            let root_norm = normalize(root);
            let mut joined = root_norm.clone();
            joined.extend(normalize_relative(path));
            let renorm = renormalize_after_join(&joined);
            if within(&root_norm, &renorm) {
                return (Decision::Allow, ReasonCode::Ok);
            }
        }
        return (Decision::Deny, ReasonCode::SandboxViolation);
    };

    for root in allowlist {
        let root_norm = normalize(root);
        if within(&root_norm, &candidate_abs) {
            return (Decision::Allow, ReasonCode::Ok);
        }
    }
    (Decision::Deny, ReasonCode::SandboxViolation)
}

/// Normalize a relative path's segments WITHOUT collapsing leading `..` against
/// nothing — we want the raw component list so it can be appended to a root and
/// re-normalized as a whole.
fn normalize_relative(path: &str) -> Vec<String> {
    path.split('/')
        .filter(|s| !s.is_empty() && *s != ".")
        .map(|s| s.to_string())
        .collect()
}

/// Re-run lexical `..` resolution over an already-joined component list (root +
/// relative tail), so a `..` in the tail can pop a real root segment and, if it
/// pops past the root, surface as a leading `..` (escape).
fn renormalize_after_join(parts: &[String]) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    for seg in parts {
        match seg.as_str() {
            "." | "" => {}
            ".." => {
                if matches!(out.last().map(String::as_str), Some(s) if s != "..") {
                    out.pop();
                } else {
                    out.push("..".to_string());
                }
            }
            other => out.push(other.to_string()),
        }
    }
    out
}

/// Apply the taint authority rule to a command's output (§28.5). A command's
/// `stdout`/`stderr` carry the `untrusted_tool_result` label; if such output is
/// being used to *authorize* the next command, it is denied
/// (`AUTHORITY_FROM_TAINTED_DATA`). Using it merely as data is fine. Delegates to
/// the shared [`crate::taint`] engine so there is one authority rule.
pub fn command_authority(output_label: &str, authorizes: bool) -> (Decision, ReasonCode) {
    match taint::check_authority(output_label, authorizes) {
        taint::TaintDecision::Allow => (Decision::Allow, ReasonCode::Ok),
        taint::TaintDecision::Deny(code) => (
            Decision::Deny,
            ReasonCode::from_str(code).unwrap_or(ReasonCode::AuthorityFromTaintedData),
        ),
    }
}

/// Result of a no-shell execution (§28.1, §28.6): the resolved argv as actually
/// passed, the process exit code, and captured stdout/stderr.
#[derive(Debug, Clone)]
pub struct ExecResult {
    pub binary: String,
    pub argv: Vec<String>,
    pub exit_code: Option<i32>,
    pub stdout: Vec<u8>,
    pub stderr: Vec<u8>,
}

/// Execute a resolved argv array by directly spawning `binary` with the argv as
/// an **array** — never through a shell (§28.1). There is no `/bin/sh -c`,
/// `cmd /c`, or PowerShell wrapper, and no interpolation/globbing/word-splitting
/// of argument values: each element of `argv` is one OS-level argument.
///
/// `argv` is the resolved argument array AFTER the binary token (i.e. the
/// program's arguments). The metacharacter defense is structural: a value like
/// `"; rm -rf / #"` is delivered to the program as a single argument string.
///
/// Returns an IO error only if the process could not be spawned; a non-zero exit
/// is a *result*, not an error (§28.6).
pub fn run_argv(binary: &str, argv: &[String]) -> std::io::Result<ExecResult> {
    // The single security-critical line: Command::new(binary).args(argv).
    // No shell, ever.
    let output = std::process::Command::new(binary).args(argv).output()?;
    Ok(ExecResult {
        binary: binary.to_string(),
        argv: argv.to_vec(),
        exit_code: output.status.code(),
        stdout: output.stdout,
        stderr: output.stderr,
    })
}
